sympy/polys/partfrac.py

```diff
@@ -116,6 +116,16 @@ def apart(f, x=None, full=False, **options):
                     pass
             return f.xreplace(dict(reps))
 
+    # Factorization over ZZ_I or QQ_I is not supported so we use the
+    # factorization routines for AlgebraicField instead.
+    dom = P.domain
+    if dom.is_ZZ_I or dom.is_QQ_I:
+        if dom.is_ZZ_I:
+            dom = dom.get_field()
+        dom_alg = dom.as_AlgebraicField()
+        P = P.per(P.rep.convert(dom_alg))
+        Q = Q.per(Q.rep.convert(dom_alg))
```

Comment:
Factorisation over `ZZ_I` and `QQ_I` does not work so I've prevented them from being used in `apart` here. Instead `QQ<I>` will be used.
```julia
In [16]: p1 = Poly(x**2 + 1, x)                                                                                                                

In [17]: p2 = Poly(x**2 + 1, x, domain=QQ_I.as_AlgebraicField())                                                                               

In [18]: p1.factor_list()                                                                                                                      
Out[18]: (1, [(Poly(x**2 + 1, x, domain='ZZ'), 1)])

In [19]: p2.factor_list()                                                                                                                      
Out[19]: (1, [(Poly(x - I, x, domain='QQ<I>'), 1), (Poly(x + I, x, domain='QQ<I>'), 1)])
```
Would it be better to make factorisation over `QQ_I` and `ZZ_I` work?

I wonder if it's possible to make `dup/dmp` treat `QQ_I` as an algebraic field when factorising. I guess that factorising over `ZZ_I` is harder...

Reply:
> Would it be better to make factorisation over `QQ_I` and `ZZ_I` work?

I think that `QQ_I` should be replaced by `QQ<I>` somewhere, maybe in `dup/dmp_ext_factor`.

Reply:
Would it be possible to add attributes or methods to `QQ_I` so that `dup/dmp` can treat it transparently as an algebraic field without needing any conversion?

Would it make sense to have `QQ_I.is_AlgebraicField` or some other attribute that is common between `QQ_I` and the `AlgebraicField` domains?

It looks like the main attributes needed for an algebraic field are `ext` and `mod`...

Reply:
It looks like at least new versions of `dup/dmp_sqf_norm` would be needed. `QQ_I` should have `conjugate` and the norm of `f` will be `f*cf` where `cf` is `f` with all coefficients conjugated.

Reply:
Would it make sense for `ZZ_I` to factorise like `QQ_I` or perhaps like `ZZ[I]`?

On master factor can factorise in some cases by treating `I` as a generator:
```julia
In [14]: e = expand((x - I)*(x + 1))                                                                                                           

In [15]: e                                                                                                                                     
Out[15]: 
 2              
x  + x - ⅈ⋅x - ⅈ

In [16]: factor(e)                                                                                                                             
Out[16]: (x + 1)⋅(x - ⅈ)
```
The PR currently gives:
```julia
In [14]: factor(e)                                                                                                                             
Out[14]: 
 2                
x  + x⋅(1 - ⅈ) - ⅈ
```

Reply:
Factorizing polynomials over `ZZ_I` should be done in two steps, factorizing the primitive part over `QQ_I` and the content in `ZZ_I`.

Reply:
What would `factor_list` return in that case?

At the moment `factor_list` returns a tuple of the content and then a list of factors of the primitive:
```julia
In [21]: Poly(x**2/2 - 2, x).factor_list()                                                                                                     
Out[21]: (1/2, [(Poly(x - 2, x, domain='QQ'), 1), (Poly(x + 2, x, domain='QQ'), 1)])
```
If the content itself factorises where would that go in this data structure?

Is there an implementation for factorisation of gaussian integers anywhere?

Perhaps if not then for now `factor_list` can return the content unfactorised...

Reply:
> Is there an implementation for factorisation of gaussian integers anywhere?

No, but that would not be hard to implement if desired. First factorize the norm in `ZZ` and then factor those prime factors that split in `ZZ_I`.